Integrate updated OP consumer

MartijnBogaert opened the pull request:
OPH-331

## Overview

OP data is being consumed by delta consumer version 0.1.0, with mappings defined by SPARQL CONSTRUCT queries.

This PR encapsulates three main updates:
- Migrations added for getting rid of data consumed by 'old' OP consumer
- Mappings added as SPARQL CONSTRUCT queries to consume OP data
- `update-bestuurseenheid-mock-login-service` added to dynamically generate a mock user for every bestuurseenheid

## How to run

1. Add the following to `docker-compose.override.yml`

```yml
services:
  op-public-consumer:
    environment:
      DCR_LANDING_ZONE_DATABASE: "virtuoso" # Only on first run
      DCR_REMAPPING_DATABASE: "virtuoso" # Only on first run
      DCR_DISABLE_INITIAL_SYNC: "false" # Only on first run
      DCR_DISABLE_DELTA_INGEST: "true" # Only on first run
  mock-bestuurseenheid-generator:
    environment:
      MU_SPARQL_ENDPOINT: "http://virtuoso:8890/sparql" # Only on first run
```

2. Run and wait for the migrations to finish

```bash
docker compose -f docker-compose.yml -f docker-compose.dev.yml -f docker-compose.override.yml up -d migrations
```

3. Run and wait for the OP consumer to finish

```bash
docker compose -f docker-compose.yml -f docker-compose.dev.yml -f docker-compose.override.yml up -d database op-public-consumer
```

4. Run and wait for the mock users to be generated

```bash
 docker compose -f docker-compose.yml -f docker-compose.dev.yml -f docker-compose.override.yml up -d mock-bestuurseenheid-generator
```

5. Update `docker-compose.override.yml`

```yml
services:
  # op-public-consumer:
  #   environment:
  #     DCR_LANDING_ZONE_DATABASE: "virtuoso" # Only on first run
  #     DCR_REMAPPING_DATABASE: "virtuoso" # Only on first run
  #     DCR_DISABLE_INITIAL_SYNC: "false" # Only on first run
  #     DCR_DISABLE_DELTA_INGEST: "true" # Only on first run
  # mock-bestuurseenheid-generator:
  #   environment:
  #     MU_SPARQL_ENDPOINT: "http://virtuoso:8890/sparql" # Only on first run
```

6. Start the (complete) application

```bash
docker compose -f docker-compose.yml -f docker-compose.dev.yml -f docker-compose.override.yml up -d
```

claire-lovisa commented on config/migrations/2025/old-consumer-flush/20250219160502-remove-old-jobs.sparql:
```diff
@@ -0,0 +1,34 @@
+
+# DELTA SYNC JOBS
+
+DELETE WHERE {
```
The query seemed heavy when I ran it on dev data, might be worth checking on prod data that it runs, otherwise you'll have to split it a bit (like jobs before a date / after that date)

MartijnBogaert replied:
Good call! I checked and the impacted triples are indeed very abundant. I chunked the query by month now.

claire-lovisa commented on config/migrations/2025/old-consumer-flush/20250220161209-remove-triples-public-graph.sparql:
```diff
@@ -0,0 +1,137 @@
+  }
+} ;
+
+DELETE WHERE {
+  GRAPH <http://mu.semte.ch/graphs/public> {
+    ?s a foaf:OnlineAccount ;
+      ?p ?o .
+  }
+} ;
+
+DELETE WHERE {
+  GRAPH <http://mu.semte.ch/graphs/public> {
+    ?s a foaf:Person ;
+      ?p ?o .
+  }
+}
```
It's nitpicking but you already remove them in the last migration :grin: or at least they should be removed if you also flush the triples without a type

MartijnBogaert replied:
You're completely right, wasn't necessary at all, I got rid of these two!

claire-lovisa commented on config/migrations/2025/old-consumer-flush/20250220163123-remove-identifiers.sparql:
```diff
@@ -0,0 +1,7 @@
+PREFIX dct: <http://purl.org/dc/terms/>
+
+DELETE WHERE {
+  GRAPH <http://mu.semte.ch/graphs/public> {
+    ?s dct:identifier ?o .
+  }
+}
```
This should be gone with `20250220161209-remove-triples-public-graph.sparql` no ?

claire-lovisa commented on config/migrations/2025/old-consumer-flush/20250220161209-remove-triples-public-graph.sparql:
```diff
@@ -0,0 +1,137 @@
+
+DELETE WHERE {
+  GRAPH <http://mu.semte.ch/graphs/public> {
+    ?s a besluit:Bestuurseenheid ;
```
You also need to explicitely flush `http://data.lblod.info/vocabularies/erediensten/BestuurVanDeEredienst`, `http://data.lblod.info/vocabularies/erediensten/CentraalBestuurVanDeEredienst` and `http://www.w3.org/ns/regorg#RegisteredOrganization`, they should inherit from the `besluit:Bestuurseenheid` type but the migration service isn't smart enough to apply that for now.

That said, after looking into the data, it seems for some reason that most of them have lost their types. Is it on the radar? Maybe now is a good time to flush them too, otherwise you might end up with double labels for example?
I found them via the following query:
```
SELECT DISTINCT ?s
WHERE {
  GRAPH <http://mu.semte.ch/graphs/public> {
    ?s ?p ?o .
    FILTER NOT EXISTS { ?s a ?type }
  }
}
```

MartijnBogaert replied:
`http://data.lblod.info/vocabularies/erediensten/BestuurVanDeEredienst`, `http://data.lblod.info/vocabularies/erediensten/CentraalBestuurVanDeEredienst` and `http://www.w3.org/ns/regorg#RegisteredOrganization` are now all taken care of at the end of the file.

MartijnBogaert replied:
Also, I added an extra query to remove all typeless triples from the public graph. You're completely right in saying we don't need those, thanks for pointing out!

claire-lovisa commented on config/migrations/2025/old-consumer-flush/20250220162634-remove-auth-concepts.sparql:
```diff
@@ -0,0 +1,12 @@
+
+DELETE WHERE {
+  GRAPH <http://mu.semte.ch/graphs/public> {
+    ?concept a skos:Concept ;
```
I'm wondering if it's not deleting too much, you're getting rid of all the NIS codes (ex. `http://vocab.belgif.be/auth/refnis2019/11004`) that are not published by the producer you consume. But if you don't need them, it's probably fine

MartijnBogaert replied:
I know it's quite a harsh delete, but I'm aware what it deletes. We didn't add the impacted triples, so they probably came through somehow in an earlier op consumer version. In any case, we don't need the concepts without a conceptscheme so they can be safely deleted.

claire-lovisa reviewed (changes requested):
Overall the consumer integration looks great, I did find some caveats in the flushing that might result in errors after the new sync is initialized. Can you check if it's relevant, as someone with more context on the projet than me?
I'll do a full test once you checked it out, thanks!

nbittich commented on manual-query-reasoning-service.sparql:
```diff
@@ -1,69 +0,0 @@
-# this query must run after the initial sync finishes
```
@claire-lovisa for some reason we couldn't make the reasoner works (the reasoner service was crashing on oph), but maybe with the new version of the consumer this has been fixed?

nbittich replied:
the error was:
```
reasoner_1       | ** ERROR ** gre ** error(existence_error(url,http://virtuoso:8890/sparql?default-graph-uri=&query=CONSTRUCT+%7B%0D%0A%3Fs+%3Fp+%3Fo%0D%0A%7D+WHERE+%7B%0D%0A+GRAPH+<http://mu.semte.ch/graphs/landing-zone/op>+%7B%0D%0A%3Fs+%3Fp+%3Fo%0D%0A%7D%0D%0A%7D&should-sponge=&format=text%2Fplain&timeout=0&run=+Run+Query),context(_5170,status(500,SPARQL Request Failed)))
reasoner_1       |
reasoner_1       | 500 op2clb/main

claire-lovisa replied:
@nbittich The new version of the consumer completely gets rid of the reasoner, problem solved haha

MartijnBogaert replied:
@claire-lovisa woops forgot to remove that, thanks for noticing!

MartijnBogaert commented on docker-compose.yml:
```diff
@@ -170,37 +170,27 @@ services:
-  op-consumer:
-    image: lblod/delta-consumer:latest
+  op-public-consumer:
+    image: lblod/delta-consumer:0.1.0
```
Good call, thanks!

MartijnBogaert commented on docker-compose.dev.yml:
```diff
@@ -11,10 +11,19 @@ services:
     links:
       - database:database
     restart: "no"
+  mock-bestuurseenheid-generator:
```
This is intentional, just to make sure we do'nt fill our production db with mock users. But you're right, I musn't forget to put this in override on the dev server 🙂

claire-lovisa reviewed (changes requested):
After some tweaking of the flushing migration I was able to test the full sync. The initial sync completed successfully but then I got 500 errors while doing the delta sync. Putting the query in virtuoso to try to debug, I saw that this error was popping up in the virtuoso sparql endpoint: `Virtuoso 42000 Error The estimated execution time 4265 (sec) exceeds the limit of 400 (sec).`. This hints for something to improve in Virtuoso's configuration. I looked it up and found that in OPH's `virtuoso.ini` the following parameter is defined:
```
MaxQueryCostEstimationTime 	= 400	; in seconds
```
I then compared to what's the value that is set in other apps and it seems to be `4000`. 

But even when bumping it to 10 000, it still wasn't coming through and the query were really slow when they are a lot faster on the other applications. I tried linking the consumer to virtuoso directly instead of the database for delta sync and there it went through. Maybe something to do with the sparql parser as database..? I hope people on the chat will have leads on what the potential issue could be. If you need help recreating the issue locally, ping me and I'll provide a more detailed "how did i get into this situation" guide!

PS: on the bright side, when wired to virtuoso directly, I was able to update the name of an admin unit in OP and it got updated in OPH :grin:

MartijnBogaert commented on config/migrations/2025/old-consumer-flush/20250219160502-remove-old-jobs.sparql:
```diff
@@ -0,0 +1,280 @@
+PREFIX cogs: <http://vocab.deri.ie/cogs#>
```
The issue didn't arise with my own triplestore, but on a copy of the production triplestore the migrations failed as well. I split up the migrations, which fixed the issue for me. How are the updated migrations behaving for you?

MartijnBogaert commented:
> After some tweaking of the flushing migration I was able to test the full sync. The initial sync completed successfully but then I got 500 errors while doing the delta sync. Putting the query in virtuoso to try to debug, I saw that this error was popping up in the virtuoso sparql endpoint: `Virtuoso 42000 Error The estimated execution time 4265 (sec) exceeds the limit of 400 (sec).`. This hints for something to improve in Virtuoso's configuration. I looked it up and found that in OPH's `virtuoso.ini` the following parameter is defined:
> 
> ```
> MaxQueryCostEstimationTime 	= 400	; in seconds
> ```
> 
> I then compared to what's the value that is set in other apps and it seems to be `4000`.
> 
> But even when bumping it to 10 000, it still wasn't coming through and the query were really slow when they are a lot faster on the other applications. I tried linking the consumer to virtuoso directly instead of the database for delta sync and there it went through. Maybe something to do with the sparql parser as database..? I hope people on the chat will have leads on what the potential issue could be. If you need help recreating the issue locally, ping me and I'll provide a more detailed "how did i get into this situation" guide!
> 
> PS: on the bright side, when wired to virtuoso directly, I was able to update the name of an admin unit in OP and it got updated in OPH 😁

A few things have been updated:
- Virtuoso is on a new version (this is already deployed on dev, so you can copy the db files from there)
- In `virtuoso.ini` I set `MaxQueryCostEstimationTime 	= 4000` and `MaxQueryExecutionTime      	= 1200`, comparable to what most LBLOD apps are set to
- I updated the `sparql-parser` from 0.0.13 to 0.0.14 (from the release notes I wouldn't expect this to make a difference for our case, but you never know)

I haven't encountered any issues yet, but then again I haven't yet received any delta updates, so I can't say for sure it does or does not work at this point. Could you check again? And if you still encouter the same issue, can you help me in recreating it? Thanks!

claire-lovisa commented:
@MartijnBogaert I just tested and it seems the issue is still there. You can try it out yourself by wiring your local OPH stack to the dev environemnt of OP for the initial sync like so:
```
services:
  op-public-consumer:
    environment:
      DCR_SYNC_BASE_URL: "https://dev.organisaties.abb.lblod.info" # or another endpoint
      DCR_LANDING_ZONE_DATABASE: "virtuoso" # for the initial sync, we go directly to virtuoso
      DCR_REMAPPING_DATABASE: "virtuoso" # for the initial sync, we go directly to virtuoso
      DCR_DISABLE_INITIAL_SYNC: "false"
      DCR_DISABLE_DELTA_INGEST: "true"
```
This executes well on my end, but then when I activate the delta sync
```
services:
  op-public-consumer:
    environment:
      DCR_SYNC_BASE_URL: "https://dev.organisaties.abb.lblod.info" # or another endpoint
      DCR_DISABLE_INITIAL_SYNC: "false"
      DCR_DISABLE_DELTA_INGEST: "false"
```
It tries to consume a change I made on OP dev (on the name of AGB Ronse, but you can change any name, it should create a delta that your stack will then try to consume). You'll see it fails to ingest it because in the logs you'll see
```
op-public-consumer-1  | Retry allowed? true
op-public-consumer-1  | Sleeping 98716 ms before next attempt
```

Can you reproduce the issue as well?

MartijnBogaert commented:
@claire-lovisa Thanks for the instructions, I managed to recreate the issue.

While trying to debug, I noticed the mock users generation service was putting a lot of strain on the sparql parser as it spits out a huge amount of queries over a period of about 2 minutes. The OP consumer firing its own queries at the same time, didn't help of course. That's why I updated the README instructions to generate the mock users after the consumer's initial sync and before starting the full application, with its queries passed to virtuoso directly.

Then the issue at hand: it seemed virtuoso wouldn't execute the following query, instead giving the error `Virtuoso 42000 Error The estimated execution time 20359 (sec) exceeds the limit of 4000 (sec).`:
```sparql
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
PREFIX organisatie: <http://lblod.data.gift/vocabularies/organisatie/>
PREFIX schema: <http://schema.org/>
PREFIX foaf: <http://xmlns.com/foaf/0.1/>
DELETE { GRAPH <http://mu.semte.ch/graphs/public> { ?s ?p ?o. } }
WHERE {
  GRAPH <http://mu.semte.ch/graphs/op/landing> {
    ?s rdf:type organisatie:TypeVestiging;
      ?p ?o.
    FILTER(?p IN(rdf:type, mu:uuid, schema:email, schema:telephone, schema:contactType, foaf:page))
  }
  VALUES ?s {
    <http://data.lblod.info/id/bestuurseenheden/ce333adebd6614199af610e487ba6a9c4c2ec6781e354b1b925d790572905efa>
  }
  VALUES ?p {
    <http://www.w3.org/2004/02/skos/core#prefLabel>
  }
  VALUES ?o {
    "AGB Ronse"
  }
}
```

I don't know exactly why it estimates this query to take such a long time, but I can imagine it has to do with the fact that our public graph has a huge amount of triples. This is an issue that we actually already noticed and in the near future we'll be moving large amounts of data to separate graphs to relieve our public graph.

For now, though, I set `virtuoso.ini`'s `MaxQueryCostEstimationTime` to `30000`, which fixed the issue for me. I know it's a very large number, but do you think it would be okay to have it set like this until we've cleaned up our public graph?

claire-lovisa commented:
@MartijnBogaert Happy to read you could reproduce it as well and already found some improvement points! I'm not sure if setting such a high value is OK or not, could you ask in rocketchat, probably in the #mu-semtech channel? People with more knowledge on virtuoso's config might have valuable insights on this.

In any case, I would put a big `TODO` in the `virtuoso.ini`'s comments next to the value linking to the work splitting the graphs to make sure not to forget to turn it down once it's ready.

MartijnBogaert commented:
@claire-lovisa I asked around a bit for the high value in `virtuoso.ini` and indeed, it's not optimal at all.

That's why I took another look and found out that the queries virtuoso trips over, are rather 'strange'. In fact, their `VALUES ?p` statements target properties that will never be available as the `FILTER` in de `WHERE` clause only allows other properties (e.g. https://github.com/lblod/app-openproceshuis/pull/53#issuecomment-2733067694). In other words, all my mapping queries are being used for every delta, as their `WHERE` clauses all have the `?s ?p ?o` pattern, and their `FILTER` statements are not being taken into account when the consumer service is figuring out which mapping queries should be transformed into `DELETE` or `INSERT` queries. This leads to these queries that don't really make sense, and virtuoso estimating it would take a huge amount of time to process them.

Going through the docs again, I noticed this is actually mentioned: the `WHERE` clause should be as straightforward as possible. That's why I refactored the mapping queries: they don't contain any filters anymore, but rather search for triple patterns with the properties explicitely defined (so no `?s ?p ?o` patterns anymore). Also, since some properties aren't always available for all resources of a certain type, I gave those their own mapping query.

With these updated mapping queries, things finally worked for me. The initial sync works properly, delta updates are executed without erros and the virtuoso config can retain its 'normal' values. What do you think?

claire-lovisa commented on config/op-consumer/mapping/queries/op2dl/pass/bestuurseenheid/bestuurseenheid.sparql:
```diff
@@ -0,0 +1,19 @@
+CONSTRUCT {
+  ?s rdf:type ?type ;
+     mu:uuid ?uuid ;
+     skos:prefLabel ?label ;
```
I'm glad you found a way of getting it to work!! I am just a little scared with the files with multiple lines in the `CONSTRUCT` because for example if you receive an update on a `skos:prefLabel` of an administrative unit, it might happen that you first receive the delete and a little while after the insert. In that case, first getting a delete on the label will cause this mapping query to also delete the type, uuid and classification of the admin unit. Then a little while after you'll receive the insert on the new label, but as the type etc will have been deleted, the label will never make it through. Might be worth decomposing them as well?

MartijnBogaert replied:
Good call! I took the opportunity and decomposed all queries to make sure all predicates are handled seperately.

claire-lovisa reviewed (changes requested):
I left a comment on `bestuurseenheid.sparql` but I think it also applies to `bestuurseenheid-classification.sparql` and `site.sparql`.

Really happy you found a way to make it work though :dancers:

MartijnBogaert commented:
> I was able to test the full flow and unable to break it which is a good news 😁

That's so good to hear, thanks for all your help with this!!
